### ISSUE TITLE
Implement Layer Normalization

### ISSUE THREAD

**Issue by rexlaw412** [deeplearning4j/deeplearning4j#4829]
#### Issue Description

Do Layer Normalization provided by DL4j?
Or any replacement of it?

#### Version Information

Please indicate relevant versions, including, if relevant:

* Deeplearning4j version : 0.9.1

**Comment by AlexDBlack** [comment]
Layer normalization (as in https://arxiv.org/abs/1607.06450 ) is not yet supported in DL4J.
I'll mark this as a feature request / enhancement - not sure if/when we'd get to it implementing it - but would love a contribution/pull request, if you are up to the challange.

**Comment by akshaysharma096** [comment]
I would like to contribute.

**Comment by AlexDBlack** [comment]
@akshaysharma096 Great! This would be implemented as a layer. I'd suggest starting with the custom layer example in the examples repo.
Basically you would need to implement:
1. Layer (configuration) class - should be very simple here
2. Layer (implementation) class - forward and backward pass methods
3. Gradient checks and other unit tests, to verify correctness

For layer config, see https://github.com/deeplearning4j/deeplearning4j/tree/master/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/layers

For layer implementation, see https://github.com/deeplearning4j/deeplearning4j/tree/master/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers

Feel free to ask if you have any questions.

**Comment by akshaysharma096** [comment]
Sure, I will do.

And for reference I will be using this paper: https://arxiv.org/pdf/1607.06450.pdf

@AlexDBlack

**Comment by AlexDBlack** [comment]
@akshaysharma096 great, looking forward to it :)

**Comment by akshaysharma096** [comment]
@AlexDBlack do we need an option to lock the gain in the implementation for Layer Normalization ?

**Comment by akshaysharma096** [comment]
Thanks @AlexDBlack

**Comment by akshaysharma096** [comment]
@AlexDBlack, I was looking at the backprop code for batch normalization, https://github.com/deeplearning4j/deeplearning4j/blob/master/deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/normalization/BatchNormalization.java

for layer normalization, I believe the implementation will almost be same except the normalizing factor, where instead of the batch size we will use the number of activations.

Comments ?

**Comment by AlexDBlack** [comment]
It will be very similar, yes.
Re-reading the paper: 
> The difference between Eq. (2) and Eq. (3) is that under layer normalization, all the hidden units in a layer share the same normalization terms µ and σ, but different training cases have different normalization terms.

So µ and σ are calculated on a per-example basis - instead of on a per-neuron basis in batch norm. (And my previous comment regarding "update global stats at training time, use global stats at test time" would be incorrect).

One thing I'm not sure on though is the handling of CNN layers. There, I assume that we would do the same thing as with batch norm - i.e., have learnable parameters for each channel (which is shared for all x/y locations).

**Comment by akshaysharma096** [comment]
Thanks @AlexDBlack

**Comment by akshaysharma096** [comment]
@AlexDBlack, have implemented the layer according to the paper, can you let me know where can I add the gradient checks?

**Comment by AlexDBlack** [comment]
Closed in favor of: https://github.com/deeplearning4j/deeplearning4j/issues/7175
